fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/postgresql/PostgreSqlResourceDAO.java

```diff
@@ -93,43 +90,54 @@ public Resource  insert(Resource resource, List<ExtractedParameterValue> paramet
 
         try {
             connection = this.getConnection();
-
-            this.fhirRefSequenceDAO = new FhirRefSequenceDAOImpl(connection);
-            this.parameterNameDAO = new PostgreSqlParameterNamesDAO(connection, fhirRefSequenceDAO);
-            this.codeSystemDAO = new PostgreSqlCodeSystemDAO(connection, fhirRefSequenceDAO);
-
             resourceTypeId = getResourceTypeIdFromCaches(resource.getResourceType());
             if (resourceTypeId == null) {
                 acquiredFromCache = false;
-                resourceTypeId = getOrCreateResourceType(resource.getResourceType(), connection);
+                resourceTypeId = this.readResourceTypeId(resource.getResourceType());
                 this.addResourceTypeCacheCandidate(resource.getResourceType(), resourceTypeId);
             } else {
                 acquiredFromCache = true;
             }
-
-            if (logger.isLoggable(Level.FINE)) {
-                logger.fine("resourceType=" + resource.getResourceType() + "  resourceTypeId=" + resourceTypeId +
+            if (logger.isLoggable(Level.FINER)) {
+                logger.finer("resourceType=" + resource.getResourceType() + "  resourceTypeId=" + resourceTypeId +
                          "  acquiredFromCache=" + acquiredFromCache + "  tenantDatastoreCacheName=" + ResourceTypesCache.getCacheNameForTenantDatastore());
             }
 
+            // TODO avoid the round-trip and use the configured data schema name
+            currentSchema = connection.getSchema().trim();
+            stmtString = String.format(SQL_INSERT_WITH_PARAMETERS, currentSchema);
+            stmt = connection.prepareCall(stmtString);
+            stmt.setString(1, resource.getResourceType());
+            stmt.setString(2, resource.getLogicalId());
+            stmt.setBytes(3, resource.getData());
```

Comment:
Nice the stored procedure is doing a lot of work.

Reply:
yeah